discover.py

```diff
@@ -0,0 +1,234 @@
+    supervisor.stopProcess(DNSMSAQ_PROCESS_NAME)
+
+def rewrite_dnsmasq_conf(path, conf_string):
+    temp = path + ".tmp"
+    with open(temp, 'w') as f:
+        f.write(conf_string)
+    shutil.move(temp, path)
+
+class DnsmasqRestarter(Thread):
+    def __init__(self):
+        super().__init__()
+        self.last_dnsmasq_command = ''
+
+    def run(self):
+        while True:
+            self.update_dnsmasq()
+            time.sleep(3)
+
+    def update_dnsmasq(self):
+        interface = app.config['DNS_INTERFACE']
+        if not interface:
+            if self.last_dnsmasq_command:
+                stop_dnsmasq()
+                self.last_dnsmasq_command = ''
+            return
+
+        command = get_dns_conf_string(interface)
+        if command == self.last_dnsmasq_command:
+            return
+
+        rewrite_dnsmasq_conf(app.config['DNS_CONF_FILE'], command)
+        reload_dnsmasq()
+        self.last_dnsmasq_command = command
+
+
+def get_dns_conf_string(interface):
+    ip = get_ipv4_addr(interface)
+    common_args = [
+        'domain-needed',
+        'bogus-priv',
+        'server=208.67.222.222',
+        'server=208.67.220.220',
+        'no-resolv',
+        'no-hosts',
+        'bind-interfaces',
+    ]
+    listen_args = [
+        'interface={}'.format(interface),
+        'listen-address={}'.format(ip),
+    ]
+
+    address_records = [
+        (node['hostname'], node['address'])
+        for node in nodes.get(interface, {}).values()
+        if not node['is_local']
+    ]
+    address_records.sort()
+    address_records.insert(0, (app.config['LIQUID_DOMAIN'], ip))
+    address_args = [
+        'address=/{}/{}'.format(*record)
+        for record in address_records
+    ]
+
+    args = common_args + listen_args + address_args
+    return "\n".join(args) + "\n"
+
+def dict_decode(data):
+    def decode(data):
+        if isinstance(data, bytes):
+            return data.decode("latin-1")
+        elif isinstance(data, str):
+            return data
+        return str(data)
+
+    return {
+        decode(key): decode(data[key])
+        for key in data
+    }
+
+def get_ipv4_addr(interface):
+    addrs = netifaces.ifaddresses(interface).get(netifaces.AF_INET)
+    if not addrs:
+        return None
+
+    ips = [
+        a['addr']
+        for a in addrs
+        if 'addr' in a and a.get('netmask') != HOST_ONLY_NETWORK_MASK
+    ]
+
+    if ips:
+        assert len(ips) == 1, "Only one ip per interface is supported."
+        return ips[0]
+    else:
+        return None
+
+def handle_new_service(name, interface, service_info):
+    properties = dict_decode(service_info.properties)
+    hostname = properties.get('liquid_hostname')
+    if not hostname:
+        return
+    is_local = (hostname == app.config['LIQUID_DOMAIN'])
+    if is_local:
+        address = get_ipv4_addr(interface)
+    else:
+        address = ".".join(str(x) for x in service_info.address)
```

Comment:
`address` is a sequence of 4 bytes, so x would be one byte. Those 4 bytes are the IP address encoded in network byte order. Maybe there is some function under `socket` that does that for us?

The reference is here: https://github.com/jstasiak/python-zeroconf/blob/master/zeroconf.py#L1347